File: day26/addTwoNumbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
 
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        Stack<Integer> stack1 = new Stack<>();
        Stack<Integer> stack2 = new Stack<>();
        while (l1 != null) {
            stack1.push(l1.val);
            l1 = l1.next;
        } 
        while (l2 != null) {
            stack2.push(l2.val);
            l2 = l2.next;
        }

        int carry = 0;
        ListNode res = null;

        while (!stack1.isEmpty() || !stack2.isEmpty() || carry == 1) {
            int s1 = stack1.isEmpty()? 0:stack1.pop();
            int s2 = stack2.isEmpty()? 0:stack2.pop();
            int val = s1 + s2 + carry;
            carry = val / 10;
            val %= 10;

            ListNode temp = new ListNode(val);
            temp.next = res;
            res = temp;
        }

        return res;
    }
}

File: day26/addTwoNumbers.py
'''
给你两个 非空 链表来代表两个非负整数。数字最高位位于链表开始位置。它们的每个节点只存储一位数字。将这两数相加会返回一个新的链表。
你可以假设除了数字 0 之外，这两个数字都不会以零开头。

进阶：如果输入链表不能修改该如何处理？换句话说，你不能对列表中的节点进行翻转。

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/add-two-numbers-ii
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        stack1, stack2 = [], []
        while l1:
            stack1.append(l1.val)
            l1 = l1.next
        while l2:
            stack2.append(l2.val)
            l2 = l2.next
        # 进位符
        carry = 0
        res = None
        while stack1 or stack2 or carry == 1:
            s1 = stack1.pop() if stack1 else 0
            s2 = stack2.pop() if stack2 else 0
            val = s1 + s2 + carry
            carry = val // 10
            val = val % 10
            temp = ListNode(val)
            temp.next = res
            res = temp
        return res
